src/Illuminate/Container/Container.php

```diff
@@ -195,9 +209,64 @@ public function resolved($abstract)
      */
     public function isShared($abstract)
     {
-        return isset($this->instances[$abstract]) ||
-               (isset($this->bindings[$abstract]['shared']) &&
-               $this->bindings[$abstract]['shared'] === true);
+        // The abstract was explicitly marked as shared.
+        if (isset($this->instances[$abstract])) {
+            return true;
+        }
+
+        // The abstract is explicitly bound to the container.
+        if (isset($this->bindings[$abstract])) {
+            return isset($this->bindings[$abstract]['shared'])
+                && $this->bindings[$abstract]['shared'] === true;
+        }
+
+        // Resolve global shared configuration.
+        $this->compileSharedPrefixes();
+
+        foreach ($this->sharedPrefixes as $prefix => $shared) {
+            if (0 === \strpos($abstract, $prefix)) {
+                return $shared;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Register a prefix for shared binding in the container.
+     *
+     * @param string $prefix
+     * @param bool   $shared Whether to enable service sharing
+     */
+    public function sharedByPrefix(string $prefix, bool $shared = true)
+    {
+        $this->sharedPrefixes[$prefix] = $shared;
+
+        $this->isSharedPrefixesCompiled = false;
+    }
+
+    /**
+     * Sets compiled shared prefixes.
+     *
+     * @param array $compiledSharedPrefixes
+     */
+    public function setCompiledSharedPrefixes(array $sharedPrefixes): void
+    {
+        $this->sharedPrefixes = $sharedPrefixes;
+
+        $this->isSharedPrefixesCompiled = true;
+    }
+
+    /**
+     * Gets compiled shared prefixes.
+     *
+     * @return array
+     */
+    public function getCompiledSharedPrefixes(): array
```

Comment:
Remove the return type

Reply:
@GrahamCampbell For what reason return type are not allowed?

I want to understand why.

Reply:
i'm curious about this...